shanoir-ng-datasets/src/main/java/org/shanoir/ng/dicom/web/StudyInstanceUIDHandler.java

```diff
@@ -0,0 +1,174 @@
+package org.shanoir.ng.dicom.web;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import javax.annotation.PostConstruct;
+
+import org.shanoir.ng.dataset.model.Dataset;
+import org.shanoir.ng.dataset.model.DatasetExpression;
+import org.shanoir.ng.dataset.model.DatasetExpressionFormat;
+import org.shanoir.ng.datasetacquisition.model.DatasetAcquisition;
+import org.shanoir.ng.datasetfile.DatasetFile;
+import org.shanoir.ng.examination.model.Examination;
+import org.shanoir.ng.examination.service.ExaminationService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+
+@Component
+public class StudyInstanceUIDHandler {
+	
+	private static final Logger LOG = LoggerFactory.getLogger(StudyInstanceUIDHandler.class);
+
+	private static final String WADO_URI_STUDY_UID_SERIES_UID = "studyUID=(.*?)\\&seriesUID";
+	
+	private static final String WADO_RS_STUDY_UID_SERIES_UID = "/studies/(.*?)/series/";
+
+	@Autowired
+	private ExaminationService examinationService;
+	
+	private HashMap<Long, String> examinationIdToStudyInstanceUIDCache;
+	
+	@PostConstruct
+	public void init() {
+		examinationIdToStudyInstanceUIDCache = new HashMap<Long, String>(1000);
+		LOG.info("DICOMWeb cache created: examinationIdToStudyInstanceUIDCache");
+	}
+	
+	/**
+	 * This method replaces StudyInstanceUIDs returned from the PACS with IDs
+	 * of examinations in Shanoir, in the Json returned.
+	 * 
+	 * @param root
+	 * @param examinationId
+	 * @param studyLevel
+	 */
+	public void replaceStudyInstanceUIDsWithExaminationIds(JsonNode root, Long examinationId, boolean studyLevel) {
+		if (root.isObject()) {
+			Iterator<String> fieldNames = root.fieldNames();
+			while (fieldNames.hasNext()) {
```

Comment:
Why iterating over all elements if its possible to get elements directly with root.get("0020000D");
Extract constants for these ? 

Also, 

> for (int i = 0; i < studyInstanceUIDArray.size(); i++) {
>       studyInstanceUIDArray.remove(i);
>       studyInstanceUIDArray.add(examinationId.toString());
> }
does not lead to some kind of concurrent modification ?
(isn't it risky to iterate based on the lenth of an "array" from which you add and remove an element ?)

Reply:
Hi Jean-Come, please find my answers below:
- root.get("0020000D") does only give you access to a direct child element, not the entire tree, what would be too magic; as 0020000D is a nested element I have to go recursively
- constants extracted, DONE
- for the array, I changed to insert(i,, but please see here: https://stackoverflow.com/questions/31582252/how-to-change-element-value-in-arraynode, but as I found other code examples, it seems to be the way to do

Reply:
> root.get("0020000D") does only give you access to a direct child element, not the entire tree, what would be too magic; as 0020000D is a nested element I have to go recursively

But you can test for every element
`if (root.get("0020000D") != null) { updateNode();}
`
And keep the recursive method to check at every level of the tree if  this node exists.
If the JSON is not that big, it's not really relevant, otherwise, it avoids us to iterate over all the elements of the tree.

Furthermore, if this change need to be done only in these two fields and these fields are present only once in the whole tree, you can stop the recursion with an adapted condition.

But once more I don't know if this is relevant as I never saw the structure of the json dicomWeb.

Reply:
Hi Jean-Come,
studyInstanceUIDs and retrieveURLs are multiple elements, that occur often in the json, so I can not stop after having replaced one element. I have no guarantee at which position they come and in which order. So I am not sure to understand, what you mean. Maybe, but not very probably, the depth might change too. And when I found the element I am not going deeper in the tree, as both elements are on the same depths. Maybe a screenshot helps here.
With kind regards,
Michael
![Bildschirmfoto 2022-03-30 um 10 49 25](https://user-images.githubusercontent.com/4560288/160791466-2ecf119f-fcc2-47ea-bbe1-2b3cd105b4aa.png)